system/jhtree/ctfile.hpp

```diff
@@ -197,91 +195,111 @@ class jhtree_decl CKeyHdr : public CWritableKeyNode
+class CWriteKeyHdr : public CKeyHdr, implements IWritableNode
+{
+public:
+    virtual void write(IFileIOStream *, CRC32 *crc) override;
+};
+
+// Data structures representing nodes in an index (other than the header)
+
+class jhtree_decl CNodeBase : public CInterface
 {
 protected:
     NodeHdr hdr;
     CKeyHdr *keyHdr;
-    byte keyType;
-    bool isVariable;
 private:
     offset_t fpos;
 
 public:
-    virtual void write(IFileIOStream *, CRC32 *crc) { throwUnexpected(); }
     inline offset_t getFpos() const { assertex(fpos); return fpos; }
-    inline size32_t getNumKeys() const { return hdr.numKeys; }
     inline bool isBlob() const { return hdr.nodeType == NodeBlob; }
     inline bool isMetadata() const { return hdr.nodeType == NodeMeta; }
     inline bool isBloom() const { return hdr.nodeType == NodeBloom; }
     inline bool isLeaf() const { return hdr.nodeType != NodeBranch; }       // actually is-non-branch.  Use should be reviewed.
     inline NodeType getNodeType() const { return hdr.nodeType; }
+    inline size32_t getNodeDiskSize() const { return keyHdr->getNodeSize(); }  // Retrieve size of node on disk
     const char * getNodeTypeName() const;
 public:
     CNodeBase();
-    void load(CKeyHdr *keyHdr, offset_t fpos);
     ~CNodeBase();
+    void init(CKeyHdr *keyHdr, offset_t fpos);
 };
 
 class jhtree_decl CJHTreeNode : public CNodeBase
 {
+protected:
+    size32_t expandedSize = 0;
+    char *keyBuf = nullptr;
+
+    virtual void unpack(const void *node, bool needCopy);  // MORE - should move into virtual load()...
+    static char *expandKeys(const void *src,size32_t &retsize);  // MORE - should rename
+    static void releaseMem(void *togo, size32_t size);
+    static void *allocMem(size32_t size);
+
+public:
+    ~CJHTreeNode();
+// reading methods
+    virtual void load(CKeyHdr *keyHdr, const void *rawData, offset_t pos, bool needCopy);
+    size32_t getMemSize() const { return sizeof(CJHTreeNode)+expandedSize; } // MORE - would be more accurate to make this virtual if we want to track all memory used by this node's info
+    inline offset_t getRightSib() const { return hdr.rightSib; }
+    inline offset_t getLeftSib() const { return hdr.leftSib; }
+
+    virtual void dump(FILE *out, int length, unsigned rowCount, bool raw) const;
+};
+
+class CJHSearchNode : public CJHTreeNode
+{
 protected:
     size32_t keyLen = 0;
     size32_t keyCompareLen = 0;
-    size32_t keyRecLen;
-    size32_t expandedSize;
-    char *keyBuf;
+    size32_t keyRecLen = 0;
 
-    void unpack(const void *node, bool needCopy);
-    unsigned __int64 firstSequence;
+    unsigned __int64 firstSequence = 0;
 
     inline size32_t getKeyLen() const { return keyLen; }
-    static char *expandKeys(void *src,size32_t &retsize);
-    static void releaseMem(void *togo, size32_t size);
-    static void *allocMem(size32_t size);
+    virtual void unpack(const void *node, bool needCopy);  // MORE - should move into virtual load()...
```

Comment:
trivial: could have 'override' decaration.

(and other 2 unpack overrides below)